server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -380,30 +389,63 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
         }
     }
 
+    private static boolean canPerformOperationBasedRecovery(
+        IndexMetaData indexMetaData, TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
+        DiscoveryNode replicaNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData replicaStore) {
+        if (replicaStore.isEmpty()) {
+            // a corrupted store - it won't be able to perform operation-based recovery
+            return false;
+        }
+        // If an index is closed or frozen, we can perform an operation-based recovery only if the last commit on the replica is safe and
+        // has the same operations as the last commit on the primary. Here we must ignore the peer recovery retention lease as we don't
+        // have the persisted global checkpoint from the replica to determine if the last commit is safe. However, the likelihood that
+        // the last commit unsafe is very small. It's probably okay to use the retention lease if the sequence numbers of the last commit
+        // from the primary and replica equal.
+        if (indexMetaData.getState() == IndexMetaData.State.CLOSE ||
```

Comment:
I don't see that the tests exercise this, but it'd be good if they did.

That said, for closed and frozen indices I think it's ok not to treat them specially and simply to use the retention lease if a retention lease exists. It's going to be very rare that any recovery is needed and if it is we don't really have any option but to perform a file-based recovery anyway, so we may as well do that onto the same node as before.

One issue we might face is that closed indices don't necessarily have a PRRL for every shard copy, and I think this could result in bad behaviour on closed indices where the files don't match (and there's no sync ID). We may need to defer special treatment of closed indices to a followup since I think this'll need a change to `RecoverySourceHandler` to address.